lib/usagereporter/teleport/aggregating/service.go

```diff
@@ -0,0 +1,152 @@
+	return backend.Key(userActivityReportsPrefix, startTime.Format(time.RFC3339), reportUUID.String())
+}
+
+func prepareUserActivityReport(
+	clusterName, reporterHostID []byte,
+	startTime time.Time, records []*prehogv1.UserActivityRecord,
+) (*prehogv1.UserActivityReport, error) {
+	reportUUID := uuid.New()
+	report := &prehogv1.UserActivityReport{
+		ReportUuid:     reportUUID[:],
+		ClusterName:    clusterName,
+		ReporterHostid: reporterHostID,
+		StartTime:      timestamppb.New(startTime),
+		Records:        records,
+	}
+
+	for proto.Size(report) > maxItemSize {
+		if len(report.Records) <= 1 {
+			return nil, trace.LimitExceeded("failed to marshal user activity report within size limit (this is a bug)")
+		}
+
+		report.Records = report.Records[:len(report.Records)/2]
+	}
+
+	return report, nil
+}
+
+// reportService is a [backend.Backend] wrapper that handles usage reports.
+type reportService struct {
+	b backend.Backend
+}
+
+func (r reportService) upsertUserActivityReport(ctx context.Context, report *prehogv1.UserActivityReport, ttl time.Duration) error {
+	wire, err := proto.Marshal(report)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	reportUUID, err := uuid.FromBytes(report.GetReportUuid())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	startTime := report.GetStartTime().AsTime()
+	if startTime.IsZero() {
+		return trace.BadParameter("missing start_time")
+	}
+
+	if _, err := r.b.Put(ctx, backend.Item{
+		Key:     userActivityReportKey(reportUUID, startTime),
+		Value:   wire,
+		Expires: startTime.Add(ttl),
+	}); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (r reportService) deleteUserActivityReport(ctx context.Context, report *prehogv1.UserActivityReport) error {
+	reportUUID, err := uuid.FromBytes(report.GetReportUuid())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	startTime := report.GetStartTime().AsTime()
+	if startTime.IsZero() {
+		return trace.BadParameter("missing start_time")
+	}
+
+	if err := r.b.Delete(ctx, userActivityReportKey(reportUUID, startTime)); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (r reportService) listUserActivityReports(ctx context.Context, count int) ([]*prehogv1.UserActivityReport, error) {
+	rangeStart := backend.ExactKey(userActivityReportsPrefix)
+	result, err := r.b.GetRange(ctx, rangeStart, backend.RangeEnd(rangeStart), count)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	reports := make([]*prehogv1.UserActivityReport, 0, len(result.Items))
+	for _, item := range result.Items {
+		report := &prehogv1.UserActivityReport{}
+		if err := proto.Unmarshal(item.Value, report); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		reports = append(reports, report)
+	}
+
+	return reports, nil
+}
+
+func (r reportService) createUserActivityReportsLock(ctx context.Context, ttl time.Duration) error {
+	if _, err := r.b.Create(ctx, backend.Item{
+		Key:     backend.Key(userActivityReportsLock),
+		Value:   []byte("null"),
```

Comment:
I made it a string containing time and host ID, both in the lock and in the description of the cluster alert (e206aa4).

Reply:
(I thought about using a label for the cluster alert, but it would have to come with a bunch of caveats about not actually ever using it as a label for RBAC purposes as it wouldn't be trustable, so I went with the metadata description field instead)